torch/csrc/autograd/engine.cpp

```diff
@@ -52,7 +52,7 @@ static thread_local bool checkpoint_valid = true;
 // on it in a few places (e.g. AccumulateGrad function).
 
 struct FunctionTask {
-  GraphTask* base_;
+  std::shared_ptr<GraphTask> base_;
```

Comment:
Hmm, I'm not convinced you need shared ownership here. GraphTask previously was stack allocated in the `Engine::execute` call, and I think this ownership is accurate, because GraphTask lives as long as we are actually executing the engine for a graph. I haven't read the rest of the patch yet though.

Reply:
It should still live as long we we are actually executing the engine for the graph because no one should hold any references to the GraphTask after it is done. Going back to the old thing should still work, but we'd be passing around raw pointers to the GraphTask between threads.